File: src/main/java/com/dec/decisland/item/ModCreativeModeTabs.java
package com.dec.decisland.item;

import com.dec.decisland.DecIsland;
import com.dec.decisland.block.ModBlocks;
import com.dec.decisland.item.category.Mask;
import com.dec.decisland.item.category.Material;
import com.dec.decisland.item.category.Weapon;
import net.minecraft.core.Holder;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.ItemStack;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.neoforge.registries.DeferredRegister;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

public class ModCreativeModeTabs {
    public static final DeferredRegister<CreativeModeTab> CREATIVE_MOD_TABS =
            DeferredRegister.create(Registries.CREATIVE_MODE_TAB, DecIsland.MOD_ID);
    private static final List<CreativeTabConfig> TAB_CONFIGS = new ArrayList<>();

    public static final Supplier<CreativeModeTab> DECISLAND_MATERIALS_TAB = registerTab(
            new CreativeTabConfig.Builder("decisland_materials_tab", Map.of(
                    "en_us", "DecIsland Materials",
                    "zh_cn", "DecIsland 材料"
            ))
                    .iconItem(() -> Material.STEEL_INGOT.get())
                    .build()
    );

    public static final Supplier<CreativeModeTab> DECISLAND_FOODS_TAB = registerTab(
            new CreativeTabConfig.Builder("decisland_foods_tab", Map.of(
                    "en_us", "DecIsland Foods",
                    "zh_cn", "DecIsland 食物"
            ))
                    .iconItem(() -> ModItems.A_BOWL_OF_RICE.get())
                    .build()
    );

    public static final Supplier<CreativeModeTab> DECISLAND_WEAPONS_TAB = registerTab(
            new CreativeTabConfig.Builder("decisland_weapons_tab", Map.of(
                    "en_us", "DecIsland Weapons",
                    "zh_cn", "DecIsland 武器"
            ))
                    .iconItem(() -> Weapon.ABSOLUTE_ZERO.get())
                    .build()
    );

    public static final Supplier<CreativeModeTab> DECISLAND_MISC_TAB = registerTab(
            new CreativeTabConfig.Builder("decisland_misc_tab", Map.of(
                    "en_us", "DecIsland Misc",
                    "zh_cn", "DecIsland 杂物"
            ))
                    .iconItem(() -> ModItems.AUTUMN_LEAVES.get())
                    .build()
    );

    public static final Supplier<CreativeModeTab> DECISLAND_CROPS_TAB = registerTab(
            new CreativeTabConfig.Builder("decisland_crops_tab", Map.of(
                    "en_us", "DecIsland Crops",
                    "zh_cn", "DecIsland 作物"
            ))
                    .iconItem(() -> ModItems.BIZARRE_CHILLI.get())
                    .build()
    );

    public static final Supplier<CreativeModeTab> DECISLAND_MASKS_TAB = registerTab(
            new CreativeTabConfig.Builder("decisland_masks_tab", Map.of(
                    "en_us", "DecIsland Masks",
                    "zh_cn", "DecIsland 面具"
            ))
                    .iconItem(() -> Mask.FRANK_MASK.get())
                    .build()
    );

    /**
     * 将指定物品栏的物品添加到输出列表
     *
     * @param tabConfig 物品栏配置
     * @param output    输出列表
     */
    private static void addItemsToTab(CreativeTabConfig tabConfig, CreativeModeTab.Output output) {
        ModItems.getItemConfigs().forEach(config -> {
            // 检查物品配置中的物品栏是否匹配
            if (config.creativeTab != null &&
                    config.creativeTab.get() == CREATIVE_MOD_TABS.getEntries().stream()
                            .filter(entry -> entry.getId().getPath().equals(tabConfig.name))
                            .findFirst()
                            .map(Holder::value)
                            .orElse(null)) {
                // 根据物品名称查找对应的物品
                ModItems.ITEMS.getEntries().stream()
                        .filter(entry -> entry.getId().getPath().equals(config.name))
                        .map(Holder::value)
                        .findFirst()
                        .ifPresent(output::accept);
            }
        });
        // 添加方块
        ModBlocks.getBlockConfigs().forEach(config -> {
            // 检查物品配置中的物品栏是否匹配
            if (config.creativeTab != null &&
                    config.creativeTab.get() == CREATIVE_MOD_TABS.getEntries().stream()
                            .filter(entry -> entry.getId().getPath().equals(tabConfig.name))
                            .findFirst()
                            .map(Holder::value)
                            .orElse(null)) {
                // 根据物品名称查找对应的物品
                ModBlocks.BLOCKS.getEntries().stream()
                        .filter(entry -> entry.getId().getPath().equals(config.name))
                        .map(Holder::value)
                        .findFirst()
                        .ifPresent(output::accept);
            }
        });
    }

    public static void register(IEventBus eventBus) {
        CREATIVE_MOD_TABS.register(eventBus);
    }

    public static Supplier<CreativeModeTab> registerTab(CreativeTabConfig config) {
        TAB_CONFIGS.add(config);
        return CREATIVE_MOD_TABS.register(config.name, () -> CreativeModeTab.builder()
                .icon(() -> new ItemStack(config.iconItem.get()))
                .title(Component.translatable("itemGroup." + config.name))
                .displayItems((parameters, output) -> {
                    // 自动添加所有属于此物品栏的物品
                    addItemsToTab(config, output);
                }).build()
        );
    }

    public static List<CreativeTabConfig> getTabConfigs() {
        return TAB_CONFIGS;
    }
}

File: src/main/java/com/dec/decisland/item/ModCreativeModeTabs.kt
package com.dec.decisland.item

import com.dec.decisland.DecIsland
import com.dec.decisland.block.ModBlocks
import com.dec.decisland.item.category.Accessory
import com.dec.decisland.item.category.Mask
import com.dec.decisland.item.category.Material
import com.dec.decisland.item.category.Weapon
import net.minecraft.core.registries.Registries
import net.minecraft.network.chat.Component
import net.minecraft.world.item.CreativeModeTab
import net.minecraft.world.item.ItemStack
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Supplier

object ModCreativeModeTabs {
    @JvmField
    val CREATIVE_MOD_TABS: DeferredRegister<CreativeModeTab> =
        DeferredRegister.create(Registries.CREATIVE_MODE_TAB, DecIsland.MOD_ID)

    private val tabConfigs = mutableListOf<CreativeTabConfig>()

    @JvmField
    val DECISLAND_MATERIALS_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_materials_tab",
            mapOf("en_us" to "DecIsland Materials", "zh_cn" to "DecIsland \u6750\u6599"),
        ).iconItem { Material.STEEL_INGOT.get() }.build(),
    )

    @JvmField
    val DECISLAND_FOODS_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_foods_tab",
            mapOf("en_us" to "DecIsland Foods", "zh_cn" to "DecIsland \u98df\u7269"),
        ).iconItem { ModItems.A_BOWL_OF_RICE.get() }.build(),
    )

    @JvmField
    val DECISLAND_WEAPONS_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_weapons_tab",
            mapOf("en_us" to "DecIsland Weapons", "zh_cn" to "DecIsland \u6b66\u5668"),
        ).iconItem { Weapon.ABSOLUTE_ZERO.get() }.build(),
    )

    @JvmField
    val DECISLAND_ACCESSORIES_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_accessories_tab",
            mapOf("en_us" to "DecIsland Accessories", "zh_cn" to "DecIsland \u9970\u54c1"),
        ).iconItem { Accessory.HEART_RING.get() }.build(),
    )

    @JvmField
    val DECISLAND_MISC_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_misc_tab",
            mapOf("en_us" to "DecIsland Misc", "zh_cn" to "DecIsland \u6742\u7269"),
        ).iconItem { ModItems.AUTUMN_LEAVES.get() }.build(),
    )

    @JvmField
    val DECISLAND_CROPS_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_crops_tab",
            mapOf("en_us" to "DecIsland Crops", "zh_cn" to "DecIsland \u4f5c\u7269"),
        ).iconItem { ModItems.BIZARRE_CHILLI.get() }.build(),
    )

    @JvmField
    val DECISLAND_MASKS_TAB: Supplier<CreativeModeTab> = registerTab(
        CreativeTabConfig.Builder(
            "decisland_masks_tab",
            mapOf("en_us" to "DecIsland Masks", "zh_cn" to "DecIsland \u9762\u5177"),
        ).iconItem { Mask.FRANK_MASK.get() }.build(),
    )

    private fun addItemsToTab(tabConfig: CreativeTabConfig, output: CreativeModeTab.Output) {
        val currentTab = CREATIVE_MOD_TABS.getEntries()
            .firstOrNull { it.id.path == tabConfig.name }
            ?.value()

        ModItems.getItemConfigs().forEach { config ->
            if (config.creativeTab?.get() == currentTab) {
                ModItems.ITEMS.getEntries()
                    .firstOrNull { it.id.path == config.name }
                    ?.value()
                    ?.let(output::accept)
            }
        }

        ModBlocks.getBlockConfigs().forEach { config ->
            if (config.creativeTab?.get() == currentTab) {
                ModBlocks.BLOCKS.getEntries()
                    .firstOrNull { it.id.path == config.name }
                    ?.value()
                    ?.let(output::accept)
            }
        }
    }

    @JvmStatic
    fun register(eventBus: IEventBus) {
        CREATIVE_MOD_TABS.register(eventBus)
    }

    @JvmStatic
    fun registerTab(config: CreativeTabConfig): Supplier<CreativeModeTab> {
        tabConfigs.add(config)
        return CREATIVE_MOD_TABS.register(config.name, Supplier {
            CreativeModeTab.builder()
                .icon { ItemStack(config.iconItem!!.get()) }
                .title(Component.translatable("itemGroup.${config.name}"))
                .displayItems { _, output -> addItemsToTab(config, output) }
                .build()
        })
    }

    @JvmStatic
    fun getTabConfigs(): List<CreativeTabConfig> = tabConfigs
}
